src/lib/eslint/CustomEslintEngine.ts

```diff
@@ -0,0 +1,120 @@
+import { Catalog, RuleGroup, Rule, RuleTarget, RuleResult, RuleViolation } from '../../types';
+import {RuleEngine} from '../services/RuleEngine';
+import {CUSTOM_CONFIG, ENGINE} from '../../Constants';
+import {EslintProcessHelper, StaticDependencies} from './EslintProcessHelper';
+import { FileHandler } from '../util/FileHandler';
+import {Logger, SfdxError} from '@salesforce/core';
+
+
+export class CustomEslintEngine implements RuleEngine {
+	private dependencies: StaticDependencies;
+	private helper: EslintProcessHelper;
+	protected logger: Logger;
+
+	getEngine(): ENGINE {
+		return ENGINE.ESLINT_CUSTOM;
+	}
+
+	getName(): string {
+		return this.getEngine().valueOf();
+	}
+
+	isCustomConfigBased(): boolean {
+		return true;
+	}
+
+	async init(): Promise<void> {
+		this.logger = await Logger.child(`eslint-custom`);
+		this.dependencies = new StaticDependencies();
+		this.helper = new EslintProcessHelper();
+	}
+
+	async getTargetPatterns(): Promise<string[]> {
+		return ["**/*.js"]; // TODO: We need a different way to set target pattern. Somehow eslintrc's ignore pattern doesn't work as expected
+	}
+
+	getCatalog(): Promise<Catalog> {
+		// TODO: revisit this when adding customization to List
+		const catalog = {
+			rules: [],
+			categories: [],
+			rulesets: []
+		};
+		return Promise.resolve(catalog);
+	}
+
+	async run(ruleGroups: RuleGroup[], rules: Rule[], targets: RuleTarget[], engineOptions: Map<string, string>): Promise<RuleResult[]> {
+		// Ignoring ruleGroups, rules
+
+		if (!this.helper.isCustomRun(engineOptions)) {
```

Comment:
The logic and contract for existing #run methods is currently bifurcated, not clear, and in some cases duplicated. It seems like it would be cleanest if the DefaultRuleManager only called #run if it was valid and the engine #run commands never short circuit and return `[]`. 

An example of the duplication is that DefaultRuleManager doesn't call run in certain instances
https://github.com/forcedotcom/sfdx-scanner/blob/dev/src/lib/DefaultRuleManager.ts#L81
`if (engineRules.length > 0 && engineTargets.length > 0) {`

But the BaseEslintEngine has  short circuit that checks for the inverted logic to short circuit
`if (!targets || targets.length === 0) {`
https://github.com/forcedotcom/sfdx-scanner/blob/dev/src/lib/eslint/BaseEslintEngine.ts#L169

I think it would be clearer if the #run command is only called if it is valid and never short circuits when it is called. Perhaps we could add a `willRun` command to the RuleEngine that consolidates all of this logic into a single place and delegates the decision to the RuleEngine, but at least then the contract and flow is clearer.

Reply:
That's a great idea. Let me try it out and update code.